src/skmultiflow/anomaly_detection/half_space_trees.py

```diff
@@ -0,0 +1,548 @@
+
+    def build_tree(self, min, max, current_depth=0):
+        """
+        Builds a single half space tree.
+
+        Parameters
+        ----------
+        min: Array of floats
+            Minimum work range for every dimension.
+        max: Array of floats
+            Maximum work range for every dimension.
+        current_depth:
+            Indicates the current depth of the tree.
+        Returns
+        -------
+        HalfSpaceTreeNode:
+            root of the tree.
+        """
+        if self.max_depth == current_depth:
+            return HalfSpaceTreeNode(depth=current_depth, internal_node=False)
+        else:
+            random_dimension = random.randint(0, self.dimensions - 1)
+            p = (min[random_dimension] + max[random_dimension]) / 2.0
+            temp = max[random_dimension]
+            max[random_dimension] = p
+            left = self.build_tree(min, max, current_depth + 1)
+            max[random_dimension] = temp
+            min[random_dimension] = p
+            right = self.build_tree(min, max, current_depth + 1)
+            return HalfSpaceTreeNode(left=left, right=right, split_attribute=random_dimension,
+                                     split_value=p, depth=current_depth)
+
+    def update_mass(self, X, is_reference_window):
+        """
+        Records mass data in the tree.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+        is_reference_window: Boolean
+            True if the classifier is in the first reference window, otherwise false
+        Returns
+        -------
+
+        """
+        return self._update_mass(X, self.root, is_reference_window)
+
+    def _update_mass(self, X, node, is_reference_window):
+        """
+        Private function where actual mass data is recorded.
+
+        Parameters
+        ----------
+        X: numpy.ndarray of shape (1, n_features)
+            Instance attributes.
+        node: HalfSpaceTreeNode
+
+        is_reference_window: Boolean
+            True if the classifier is in the first reference window, otherwise false
+
+        """
+        if node is None:
+            return
+
+        if is_reference_window:
+            node.r = node.r + 1
+        else:
+            node.l = node.l + 1
+
+        if node.depth < self.max_depth:
+            if X[node.split_attribute] > node.split_value:
+                self._update_mass(X, node.right, is_reference_window)
+            else:
+                self._update_mass(X, node.left, is_reference_window)
+        else:
+            return
+
+    def update_model(self, node):
+        """
+        Updates model to adapt to new data.
+
+        Parameters
+        ----------
+        node: HalfSpaceTreeNode
+
+        """
+        if node is not None:
+            if node.r is not 0 and node.l is not 0:
+                node.r = node.l
+
+            if node.l is not 0:
+                node.l = 0
+
```

Comment:
An `if` statement is missing:
```python
if node.depth < self.max_depth:
```

Reply:
There is 

> if node is not None:

In the beginning of that code snippet, I don't know why we should add that if statement.

Reply:
This is not the same if condition, or am I missing something?

Reply:
I am a bit confused on this one. Could you a bit explain why you would like to add this `if `statement?

`if node.depth < self.max_depth:
`

Reply:
You are right, I was confused. The `if node is not None` condition has the same purpose. My bad.